### ISSUE TITLE
🌟 Epic: GitHub Sync - Amplication branch

### ISSUE THREAD

**Issue by mshidlov** [amplication/amplication#3751]
To reduce user changes overrides amplication is going to integrate with the user’s repository through a single `amplication `branch. The amplication branch will be a long-lived branch created in the user repository by amplication and will be the target location for all amplication commits. when a resource is built a corresponding commit with amplication recent changes will be created. user will be able to merge amplication-made changes by merging pull requests made by amplication from the amplication branch to the user repository master branch (a repository-specific value).

## Amplication Branch
The `amplication` branch will be a long-lived branch on the user repository, and will be created lazily when a commit is made.
when a commit is made if the repository doesn't have an `amplication` branch, then get the repository master branch and create amplication branch from the repository master branch.

## Resource commit

When a resource is built, there will be a trigger to create a commit on the changed files. (**changed files:** the changed files are all the files that were added, removed, or changed from the last build). the commit message should be `[Service Name]_[Commit Message]`. After a commit was made a check should be applied to validate that there is an amplication branch open pull request. if this is not the case, a pull request should be opened. the pull-request title should be “Recent changes from Amplication” and the description should be "[TBD]".

## Triggering a commit

As part of the build flow, amplication-server performs create pull request step. 

1. Get the resource connected repository - if the resource is not synced with GitHub the process stops here.
2. Create a new ActionStep for the pull request step
3. Create a new ActionStep ActionLog and log the pull request process has started message
4. Send Create git commit message through Kafka topic `git.internal.commit-initiated.request.0`

**Create git commit message:**

```tsx
interface GitCommitInitiated {
    commit: { id: string, message: string },
    build: { id: string, previousBuildId: string },
    repository: { owner: string, name: string, installationId: string }
}
```
![Untitled](https://user-images.githubusercontent.com/91742238/189112907-ac0a4f51-1cbd-44da-a36b-1ee642bade74.png)

## Creating a commit

git pull request service will consume  Create git commit messages from  `git.internal.create-commit.request.0` kafka topic. when the service receive a commit creation request:

1. Read and compare build files with the previous build files and collect all changed files (new, deleted and updated files).
2. Get amplication branch head commit - in case amplication branch is missing
    1. get the repository master branch head commit
    2. create a new amplication branch from the repository master branch head commit
3. create a commit on amplication branch that includes the changed files
4. Get amplication branch open pull-request - if there is no open pull request
    1. Get the repository master branch
    2. open a new pull request with the repository master branch as a base
5. Send git commit created message to kafka topic `git.internal.commit-created.event.0`

**Git commit created event:**

```tsx
interface GitCommitCreated {
    build: string,
    commit:{ id: string, sha: string, timestamp: string },
    pullRequest:{ created:boolean, url: string },
}
```
![Untitled (1)](https://user-images.githubusercontent.com/91742238/189113050-1a663384-f6da-4c61-9936-cad84c3f6b52.png)

## Logging and updating commit status

amplication server will consume git commit created messages from `git.internal.commit-created.event.0` Kafka topic. when amplication server will receive a commit created message:

1. Create a commit created ActionLog
2. Create a pull-request ActionLog
3. Update the resource github last synced and github last commit message
4. Update the ActionStep status

![Untitled (2)](https://user-images.githubusercontent.com/91742238/189113141-752c943a-8ba3-49d2-99fd-b0ab26365a75.png)

## Commit race condition

As described in this document after a resource is built the changed files generated in the build will be committed. In a wider aspect, resource build is triggered as part of a project commit, and in that process, all project resources are build in parallel the resources.

A git branch functions like a linked list where the branch points to the head of the list, the last commit made (this is the reason it is called head commit), and each commit has one or more parents, a pointer to the previous commits.

When a commit is created it should include the commit parent, and because resources triggers parallel commits to the same branch a race condition is in place for the branch head commit.
![Untitled (3)](https://user-images.githubusercontent.com/91742238/189113266-ca2777d5-f36a-477a-a3a7-dfa5570693c1.png)

### Solving the commits race condition

To ensure that we are not overriding commits we decided that the process of creating commits will be synchronous instead of parallel. 

To create a synchronous processing mechanism we are going to harness kafka key distribution mechanism. To do so you need to specify a key for each message and Kafka will put all messages with the same key in the same partition.

Kafka’s topics are divided into several partitions. While the topic is a logical concept in Kafka, a partition is the smallest storage unit that holds a subset of records owned by a topic. Each partition is a single log file where records are written to it in an append-only fashion.

By default, Kafka uses the key of the message to select the partition of the topic it writes to. This is done in the `DefaultPartitioner`if there is no key provided, then Kafka will partition the data in a round-robin fashion.

To ensure all repository commits are created in order and not in parallel we will use the git repository id as the message key to all git commit create messages. the git repository id is a UUID that makes it a unique and random value that will help with even partition distribution.

**Kafka Message:**

```tsx
{
  "key": "${repository-id}",
  "headers": {},
  "value": CreateGitCommit
}
```

When a consumer group consumes the partitions of a topic, Kafka makes sure that each partition is consumed by exactly one consumer in the group. Sending all messages to the same Kafka topic partition will ensure that a single pod of git pull service will consume all the repository commit messages.  

To ensure that repository commits are made
 in sync the consumer should consume messages one by one.

![Untitled (4)](https://user-images.githubusercontent.com/91742238/189113355-8e20bf87-2a71-42a4-9ea6-c165c36bf7e4.png)

*DevOps to allow maximum concurrency between repositories we need to make sure we have sufficient partitions so different repositories will have the minimum effect on each other.

# Sub tasks

- [x] #4592
- [x] #4691
- [ ] #4590
- [ ] #4593
- [x] #4594
- [ ] #4690
- [Define a Kafka strategy](https://github.com/orgs/amplication/projects/6/views/6?filterQuery=no%3Asprint+status%3ATodo+refactor+kafka)

**Comment by overbit** [comment]
Outdated information. New Epic issue is #5112
